libs/sfm/triangulate.cc

```diff
@@ -87,74 +87,93 @@ is_consistent_pose (Correspondence2D2D const& match,
+            if (this->opts.angle_threshold > 0.0)
             {
-                double const cos_angle = rays[i].dot(rays[j]);
-                smallest_cos_angle = std::min(smallest_cos_angle, cos_angle);
+                math::Vec3d camera_pos;
+                pose_pair[0]->fill_camera_pos(&camera_pos);
+                math::Vec3d ray0 = (tmp_pos - camera_pos).normalized();
+                pose_pair[1]->fill_camera_pos(&camera_pos);
+                math::Vec3d ray1 = (tmp_pos - camera_pos).normalized();
+                double const cos_angle = ray0.dot(ray1);
+                if (cos_angle > this->cos_angle_thres)
+                    continue;
             }
 
-        if (smallest_cos_angle > this->cos_angle_thres)
-        {
-            if (stats != nullptr)
-                stats->num_too_small_angle += 1;
-            return false;
-        }
-    }
+            /* Chek error in all input poses and find outliers */
+            std::vector<std::size_t> tmp_outliers;
+            for (std::size_t i = 0; i < poses.size(); ++i)
+            {
+                math::Vec3d x = poses[i]->R * tmp_pos + poses[i]->t;
+
+                /* Reject track if it appears behind the camera. */
+                if (x[2] <= 0.0)
+                {
+                    tmp_outliers.push_back(i);
+                    continue;
+                }
+
+                x = poses[i]->K * x;
+                math::Vec2d x2d(x[0] / x[2], x[1] / x[2]);
+                double error = (positions[i] - x2d).norm();
+                if (error > this->opts.error_threshold)
+                    tmp_outliers.push_back(i);
+            }
 
-    /* Compute reprojection error of the track. */
-    double average_error = 0.0;
-    for (std::size_t i = 0; i < poses.size(); ++i)
-    {
-        math::Vec3d x = poses[i]->R * *track_pos + poses[i]->t;
+            /* Select triangulation with lowest amount of outliers */
+            if (tmp_outliers.size() < best_outliers.size())
+            {
+                best_pos = tmp_pos;
+                std::swap(best_outliers, tmp_outliers);
+            }
 
-        /* Reject track if it appears behind the camera. */
-        if (x[2] < 0.0)
-        {
-            if (stats != nullptr)
-                stats->num_behind_camera += 1;
-            return false;
         }
 
-        x = poses[i]->K * x;
-        math::Vec2d x2d(x[0] / x[2], x[1] / x[2]);
-        average_error += (positions[i] - x2d).norm();
+    /* If all pairs have small angles pos will be 0 here */
+    if (best_pos.norm() == 0.0f)
+    {
+        if (stats != nullptr)
+            stats->num_too_small_angle += 1;
+        return false;
     }
-    average_error /= static_cast<double>(poses.size());
 
-    /* Reject track if the reprojection error is too large. */
-    if (average_error > this->opts.error_threshold)
+    /* Check if required number of inliers is found */
+    if (poses.size() < best_outliers.size() + this->opts.min_num_views)
```

Comment:
I guess it doesn't make a difference here as `outliers.size()` should always be smaller than `poses.size()`. I am just trying to avoid subtraction with unsigned types in general.